File: backend/DezibotDebugInterface.Api/Common/Models/Dezibot.cs
namespace DezibotDebugInterface.Api.Common.Models;

public class Dezibot
{
    public string Ip { get; }
    
    public DateTime LastConnectionUtc { get; set; }
    
    public List<Debuggable> Debuggables { get; }
    
    public List<LogEntry> Logs { get; }
    
    public Dezibot(string ip, DateTime lastConnectionUtc, List<Debuggable>? debuggables = null, List<LogEntry>? logs = null)
    {
        Ip = ip;
        LastConnectionUtc = lastConnectionUtc;
        Debuggables = debuggables ?? [];
        Logs = logs ?? [];
    }
    
    public record LogEntry(DateTime TimestampUtc, string LogLevel, string Message);

    public record Debuggable(string Name, List<Debuggable.Property> Properties)
    {
        public record Property(string Name, List<Property.TimeValue> Values)
        {
            public record TimeValue(DateTime TimestampUtc, string Value);
        }
    }
}

File: frontend/types/Dezibot.ts
class Dezibot {
    ip: string
    components: Component[] = []
    logs: LogEntry[] = [] // message and data

    constructor(ip: string) {
        this.ip = ip
    }

    static fromJson(json: any): Dezibot {

        json.components = json.classes
        delete json.classes        
        return json
    }
}

class Component {
    name: string
    properties: Property[] = []
    
    constructor(name: string) {
        this.name = name
    }
}

class Property {
    name: string
    values: {timestampUtc: string, value: string}[] = []

    constructor(name: string) {
        this.name = name
    }
}

class LogEntry { 
    timestampUtc: string
    level: string
    className: string
    message: string
    data: string

    constructor(
        timestampUtc: string,
        level: string,
        className: string,
        message: string,
        data: string,
    ) {
        this.timestampUtc = timestampUtc
        this.level = level
        this.className = className
        this.message = message
        this.data = data
    }
}

export {Dezibot, Component, LogEntry, Property}